### PR TITLE
Implement table reader protocol and unified access

### PR DESCRIPTION
**Pull request by jonatanklosko**
:cat2:

### CODE REVIEW

File: lib/table.ex [hunk 1]
```diff
@@ -1,2 +1,329 @@
...
+  of that, tabular access itself has two variants, a row-based one
+  and a column-based one, each useful under different circumstances.
+
+  The `Table` package provides a thin layer that unifies access to
+  tabular data in different formats.
+
+  ## Protocol
+
+  The unified access is enabled for structs implementing the
+  `Table.Reader` protocol. Note that the struct may be representable
```
Comment (josevalim) [inline lib/table.ex:16]:
```suggestion
  `Table.Reader` protocol. Note that a struct may be representable
```

File: lib/table.ex [hunk 2]
```diff
@@ -1,2 +1,329 @@
...
+  """
+
+  alias Table.Reader
+
+  @type column :: term()
+
+  @type table_info :: %{columns: list(column())}
+
+  @doc """
+  Accesses tabular data as a sequence of rows.
```
Comment (josevalim) [inline lib/table.ex:57]:
```suggestion
  Accesses tabular data as a sequence of rows.
  
  Returns an enumerable that emits each row as a map.
```

File: lib/table.ex [hunk 3]
```diff
@@ -1,2 +1,329 @@
...
+
+  defp build_row(columns, values, only) do
+    for {column, value} <- Enum.zip(columns, values),
+        include_column?(only, column),
+        into: %{},
+        do: {column, value}
+  end
+
+  @doc """
+  Accesses tabular data as individual columns.
```
Comment (josevalim) [inline lib/table.ex:125]:
```suggestion
  Accesses tabular data as individual columns.
  
  Returns a map with enumerables as values.
```

File: lib/table/reader/enumerable.ex [hunk 4]
```diff
@@ -0,0 +1,99 @@
+defmodule Table.Reader.Enumerable do
```
Comment (josevalim) [inline lib/table/reader/enumerable.ex:1]: If you want, you can put these functions in the Table.Reader.List module and invoke it from from Table.Reader.Map too. Your call.

**Review by josevalim — APPROVED** [review]
Minor suggestions and ship it!

File: lib/table/reader/enumerable.ex [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+  defp record_values(record, columns) when is_map(record) do
+    Enum.map(columns, fn column ->
+      case Map.fetch(record, column) do
+        {:ok, value} ->
+          value
+
+        :error ->
+          raise "map records must have the same columns, missing column #{inspect(column)} in #{inspect(record)}"
+      end
```
Comment (josevalim) [inline lib/table/reader/enumerable.ex:78]: Prefer pattern matching when possible:

    case record do
      %{^column => ...

File: lib/table/reader/enumerable.ex (hunk 4 above)
Comment (jonatanklosko) [reply lib/table/reader/enumerable.ex:1]: I was considering it, but feels a bit artificial, so will stick with the module :)

File: lib/table/reader/enumerable.ex (hunk 5 above)
Comment (jonatanklosko) [reply lib/table/reader/enumerable.ex:78]: Totally, good call!

**Pull request merged**